Add proof rules for hcom

wilcoxjay opened the pull request:
This is a work in progress on adding hcom rules to RedPRL's proof theory. These are summarized on pg 46 of "Dependent cubical realizability".

- [x] Add hcoms to the syntax view
- [x] Structural equality rule for hcoms
- [x] "Cap" rule
- [x] "Tube" rule
- [x] Figure out a better way of getting HCom.Eq rule to be called by auto
- [x] Add test for Cap rule
- [x] Add test for Tube rule

wilcoxjay commented:
Okay, I've pushed a prototype of the structural equality rule as HCom.Eq.

One issue I ran into is that in the implementation of `auto`, `StepEq` is convinced that all my `hcom`s are redexes, but no step ever makes progress. This may be related to the way the syntax view "loses information" as discussed on IRC. 

I hacked around this for testing purposes by just calling the `HCom.Eq` rule on all redexes :) But we should fix this before merging.

@jonsterling Could you give me some early feedback on this? And also do you have any thoughts about the redex problem mentioned above?

wilcoxjay commented:
Oh, and don't forget to checkout the new `test/success/bool-hcom.prl`, which contains the first nontrivial cubical theorem proved in RedPRL! 🛠

jonsterling commented:
@wilcoxjay Regarding the auto-step rule, I can make a few hopefully helpful comments:

1. This may have something to do with the information loss we discussed, but there's a more fundamental issue.

2. The extent to which "direct computation rules" are possible for Cubical Type Theory is different from in Nuprl-style type theory. As a result, we have a very (*very*) conservative approximation which says *Don't do any direct computation if the term has any free dimensions in it*. This is captured in `safeEval` or whatever it is called.

3. When we are checking whether a term is a redex, we should ideally also be checking that condition there—since we should not attempt to reduce terms that have free dimensions.

Does this help?

jonsterling commented on test/success/bool-hcom.prl:
```diff
@@ -0,0 +1,12 @@
+] by [
+  { lam a. lam b. lam c. lam d.
+    lam pab. lam pac. lam pbd.
+    <i> 'hcom{i; 0 ~> 1}(bool; ,pab @ i; {j}. ,pac @ j; {j}. ,pbd @ j)
```
wow! the first hcom!

jonsterling commented:
More generally, though, these hcoms really will always look like redexes to the machine, because `hcom{NONE,...}(bool;...)` reduces to `hcom{BOOL,...}(...)`.

jonsterling commented on src/redprl/syntax.sml:
```diff
@@ -15,6 +17,7 @@ struct
    | DFUN of 'a * variable * 'a | LAM of variable * 'a | AP of 'a * 'a
    | DPROD of 'a * variable * 'a | PAIR of 'a * 'a | FST of 'a | SND of 'a
    | ID_TY of (symbol * 'a) * 'a * 'a | ID_ABS of symbol * 'a | ID_AP of 'a * param
+   | HCOM of param list (* extents *) * dir * 'a (* type *) * 'a (* cap *) * (symbol * 'a) list (* tubes *)
```
one way to make this conversion lossless (I think) would be to include a boolean parameter here that says whether the hcom is canonical—which would only be meaningful assuming that the type parameter is a canonical type. Or maybe there is a cleaner way...

wilcoxjay replied:
@jonsterling Coming back to this, do you think making the conversion losless is important to do before merging this PR?

jonsterling replied:
I don't think it's crucial for this PR; already, the syntax view is fairly fast-and-loose. We might make cleaning it up a separate task, rather than burdening you with it right now.

wilcoxjay commented:
Okay, I added some documentation about restriction and fixed style issues. 

@jonsterling I'm still not sure how to proceed getting `auto` to call the hcom rule. As you mentioned, the ABT machinery will always see source-level hcom's as redexes, but `safeEval` will refuse to do anything with them. Say we extend `auto`'s redex check to look for free dimensions. I'm still not sure which lower level tactic should be extended with the hcom rule.

By the way, would it make sense to have `safeEval` raise an exception instead of returning its argument in the case where a free dimension is present? This would at least prevent one kind of infinite loop, caused by disagreement about whether the term can step between the machine and `safeEval`.

jonsterling commented:
I will try to answer this as soon as I can (exams on Monday, so it may take a few days). thanks again for your hard work!

jonsterling commented:
1. Somehow, I think that it should be fine to make `safeEval` raise an exception rather than returning its argument; it seems like ideally we would not be collapsing failure in the rules themselves like we are doing now. Feel free to try it—I can't guarantee that depends on the current behavior, but hopefully nothing does.

2. As for adding a call to the hcom structural equality rule, I think that hcoms should be dealt with specially at the level of the automation tactics. Therefore, maybe the right thing to do is to make `StepEq` have an outer case statement that checks for hcoms, and then delegates to the appropriate rule; if it is not an hcom, then, pass to the existing canonicity-based dispatch.

cangiuli commented:
I haven't been watching closely, but what exactly is the intent of `safeEval` right now? I think we decided that one can evaluate safely so long as the principal argument's head constructor has no dimension names as indices, no?

jonsterling commented:
@cangiuli That would be quite nice, but the last time we spoke about it, you said that we could only evaluate if the entire term had no free dimension names. Previously, it was implemented such that only the head operator's indices were checked, but I seem to remember that we discussed it and you thought that would be unsound. Perhaps we had a miscommunication, though!

wilcoxjay commented:
You two are the experts, but to my eye it looks like `safeEval` is actually only checking the head operator's indices. See [here](https://github.com/RedPRL/sml-redprl/blob/8cf6e816720cd7a8fe147a6f412d0300be015a76/src/redprl/refiner.fun#L1032)

jonsterling commented:
@wilcoxjay Huh, I must have got what happened backwards or messed up 😄  @cangiuli

jonsterling commented:
Ah, I see what `safeEval` is doing now; I should have just looked at it before I commented from memory 15 minutes ago, haha.

It checks the head operator at every *step*; this should be sound, but it may not be quite ideal. As @cangiuli mentioned, it would be better to be checking the indices of the principal argument; however, I'd have to think a bit about how to achieve that generically (it ought to be possible with the machine apparatus that's in place, just have to think about it). For now, we should probably just leave it the way it is until we propose the proper way to do it.

cangiuli commented:
@jonsterling It doesn't matter whether or not the term has dimensions internally; the indices of the principal argument are, I think, the only things that pose a problem.

Let me write `R` for the "cubically-stable stepping judgment", which I think is the intent of `safeEval`. For example, `R((\a.M) N , M[N/a])` regardless of free dimensions, since dimensions in `M`,`N` can't possibly affect this step. Also note that `R(M@x , M'@x)` when `R(M,M')`, even though `x` is the index of the head operator, because the identity of that index does not affect the computation until `M` is a value. Does this sound right?

jonsterling commented:
@cangiuli This is very very helpful!

But it sounds like it contradicts some of the things we'd discussed in the past, which is the idea of being (not) able to step an `hcom` which is non-canonical if it has a dimension variable in its indices. Can I read your comment as a confirmation that it is actually safe to step `hcom{x...}(A -> B; ...)` to `\a. hcom{x...}` or whatever?

jonsterling commented:
@cangiuli Aha, I just re-read your comment and now I understand what you were saying. So, that would be a resounding "no" to my last question 😄 

I think the relation you describe makes a lot of sense, and we should try to implement it.

cangiuli commented:
@jonsterling Right, when `hcom{x...}` is the principal argument, this is precisely the scenario in which stepping is not safe. But I think it would be prudent for us to discuss this more carefully before implementing something fancy.

jonsterling commented:
@cangiuli OK, sounds good! Soon (once I get onto vacation), I could come up with a proposal for such a relation, with respect to our machine dynamics, which you could review and we could discuss.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+       page 46.
+
+       Such automation is possible because the rules are "syntax
+       directed" on the restriction being performed.
```
great! This is just the sort of thing that Carlo and I had envisioned, and thanks for putting the reference here.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+
+    fun neqDims (P.APP P.DIM0, P.APP DIM1) = true
+      | neqDims (P.APP P.DIM1, P.APP DIM0) = true
+      | neqDims _ = false
```
What is the meaning of `neqDims`? I'm curious why it always returns false on "generic" / variable dimensions.

wilcoxjay replied:
`neqDims` is used to implement the part of the restriction judgment that "finds a contradiction" such as `0 = 1`. Since `x = 0` is not a contradiction, it is important that both sides are constants, and that they are different. I'll clarify this.

jonsterling replied:
Ah, ok! Thanks!

Maybe it could be renamed to something like `dimsContradictory` or something else?

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+          let
+            val J = H >> CJ.EQ ((substSymbol (r,u) tube, cap), ty)
+          in
+            List.map (fn j => #1 (makeGoal j)) (Restriction.One J ext eps)
```
might as well write `#1 o makeGoal`

wilcoxjay commented:
Okay, I've finished prototyping all the rules. 

I'd appreciate another pair of eyes looking at the rules on paper and making sure the code matches. 

I'd also appreciate any ideas for examples/tests that use the cap and tube rules. 

One thing I'm not sure about for the cap and tube rules is how useful they are when stated the way they are, since they require the right hand side to already be the right thing. The proof theory doesn't (yet?) have any sort of substitution or rewrite rule, so I imagine it's rather difficult to get your goal into a form where these rules apply.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+            List.map (fn j => #1 (makeGoal j)) (Restriction.One J ext eps)
+          end
+      in
+        listToTel (ListMonad.bind (fn (ext, eps, tube) => tubeCap ext eps tube) group)
```
If you change `tubeCap` to be curried, this could be written `listToTel (ListMonad.find tubeCap group)`, which might be a little cleaner.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+
+        val interTubeGoals = listToTel
+          (ListMonad.bind (fn ((ext, eps, tube0), (_, _, tube1)) => interTube ext eps tube0 tube1)
+                          (ListPair.zipEq (group0, group1)))
```
I don't really care that much (so feel free to ignore this comment), but for reference, I prefer the following indentation style (which avoids leading whitespace that depends on lexical width of other identifiers):

```sml
val interTubeGoals = 
  listToTel
    (ListMonad.bind 
      (fn ((ext, eps, tube0), (_, _, tube1)) => interTube ext eps tube0 tube1)
      (ListPair.zipEq (group0, group1)))
```

jonsterling commented:
@wilcoxjay Good point about the "judgmental computation rules" for cap & tube... What I've done in the past is taken the rules on paper and "transposed" them, so that one side of the equation doesn't have any constraints on it. Essentially, this is like building transitivity into each rule, which is a bit better for refinement proof for the reasons you describe. An example of this in practice is `Path.ApComputeConst`.

wilcoxjay commented:
@jonsterling Hmm, okay, I'll try to implement transposed versions of those rules. But I don't really understand what's going on in `Path.ApComputeConst`. What does the `Synth` judgment mean? Something like "give me a type for this term"?

If we translate the code back to an on paper rule, would it be something like (horrible mix of RedPRL and TeX syntax...)
```
    m in paths(x.A; P_0; P_1)
    A<\epsilon / x> = A' type
    P_\epsilon = P in A'
------------------------------
  m @ \epsilon = P in A'
```

jonsterling commented:
@wilcoxjay Yeah, the implementation there is a bit arcane 😄 `R synth` is indeed a judgment that means `tell me what type R is`... This is a nice trick that is enabled by dependent refinement, since when we add this subgoal, we bind a variable which ranges over whatever type `R` turns out to be; then, we can drill down into this type using the `MATCH` judgment. This gives us a way to sort of "asynchronously" depend on stuff from an question that we may not yet have fully answered.

I think your translation of the rule looks about right.

The main idea of this "transposition" is that you take an equational rule `M = N : A` and replace it with something like `M = L : A ====> N = L : A`. Then, the rule can be applied without the user having to get the goal into just the right shape.

wilcoxjay commented:
Okay, I think I got it; that was easier than I expected :) 

@jonsterling does that look right?

wilcoxjay commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1013,190 @@ struct
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.TYPE ty
+        val (goalCap, _) = makeGoal @@ H >> CJ.MEM (cap, ty)
+        val (goalEq, _) = makeGoal @@ H >> CJ.EQ ((cap, rhs), ty)
```
So I just added this goal when mechanically transposing the rule, but it looks to me like it makes the previous goal redundant. 

Isn't it the case that if `a = b in A` then `a in A`?

jonsterling commented:
Great work @wilcoxjay! I'll take some time today to review the rules and compare them with what's on paper.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1015,224 @@ struct
+        val group = groupTubes exts tubes
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.TYPE ty
+        val (goalCap, _) = makeGoal @@ H >> CJ.MEM (cap, ty)
```
As you described, let's delete this goal since it is redundant.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1015,224 @@ struct
+      end
+
+    (* Search for an index of a constant extent in an hcom, to pass to HCom.TubeEq. *)
+    fun FindTubeEq alpha jdg =
```
Good idea, factoring this out and having the primitive rule take `i` as a parameter.

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1015,224 @@ struct
+
+        val w = alpha 0
+
+        fun interTube ext eps (u, tube0) (v, tube1) =
```
Can you add a comment here which describes what this does, analogous to the comment on `intraTubeGoals`? In these comments, it might also be nice to write (in ascii) an example premise that they correspond to from the paper.

wilcoxjay replied:
Good idea, I just pushed some comments!

jonsterling commented on src/redprl/refiner.fun:
```diff
@@ -1015,8 +1015,224 @@ struct
+    fun listToTel l = List.foldl (fn (g, l) => l >: g) T.empty l
+
+    (* Produce the list of goals requiring that tube aspects agree with each other. *)
+    fun intraTubeGoals H w ty group =
```
Am I right to understand that this corresponds to the premise that says `forall i, j, eps, eps'. N_i^eps = N_j^eps' in A [Psi, y | r_i = eps, r_j = eps']` (the fourth premise of the structural equality rule)?

If so, does this do the right thing? From the rule it looks like we need to try every possible combination of tube faces or something, but I can't quite make out how that's happening here. But I might be misunderstanding either the paper rule or the implementation.

wilcoxjay replied:
Yes, that's right. We do need to consider all pairs of tube faces. The local function `intraTube` produces the (one or zero) goals corresponding to a particular choice of tube faces. Then we wrap that up in a couple `ListMonad.bind`s to get all of them. 

Does that seem right?

jonsterling replied:
Ahhhh, it's the `bind` that makes it all work 😄 Thanks for explaining!
